lib/BackendName.hs

```diff
@@ -0,0 +1,64 @@
+-- SPDX-FileCopyrightText: 2022 Serokell <https://serokell.io>
+--
+-- SPDX-License-Identifier: MPL-2.0
+
+module BackendName
+  ( BackendName
+  , backendNameCharSet
+  , newBackendName
+  , getBackendName
+  , backendNameCodec
+  ) where
+
+import Fmt (Buildable)
+import Data.Hashable (Hashable)
+import qualified Data.Aeson as A
+import qualified Data.Text as T
+import qualified Toml
+import qualified Data.HashMap.Strict as HS
+import Data.Bifunctor (Bifunctor(first))
+import Control.Monad ((>=>))
+import Validation (Validation(Failure, Success))
+
+newtype BackendName = UnsafeBackendName T.Text
+  deriving stock (Show, Eq)
+  deriving newtype (A.ToJSON, A.ToJSONKey, A.FromJSON, A.FromJSONKey, Hashable, Buildable)
+
+backendNameCharSet :: [Char]
+backendNameCharSet = ['a'..'z'] ++ ['A'..'Z'] ++ ['0'..'9'] ++ "-_;"
+
+newBackendName :: T.Text -> Either T.Text BackendName
+newBackendName t
+  | T.null t =
+      Left "Backend name should contain at least 1 character"
+  | T.any (`notElem` backendNameCharSet) t =
+      Left $ "Backend name can only contain the following characters: '" <> T.pack backendNameCharSet <> "'"
+  | otherwise =
+      Right $ UnsafeBackendName t
+
+getBackendName :: BackendName -> T.Text
+getBackendName (UnsafeBackendName t) = t
+
+_BackendName :: Toml.TomlBiMap T.Text BackendName
+_BackendName = Toml.BiMap textToBackendName backendNameToText
+  where
+    textToBackendName :: T.Text -> Either Toml.TomlBiMapError BackendName
+    textToBackendName t = first Toml.ArbitraryError (newBackendName t)
+
+    backendNameToText :: BackendName -> Either Toml.TomlBiMapError T.Text
+    backendNameToText = Right . getBackendName
+
+backendNameCodec :: Toml.Key -> Toml.TomlCodec BackendName
+backendNameCodec backendNameKey = Toml.Codec input output
```

Comment:
We can simply this implementation by reusing `Toml.text` codec, and then mapping back/forth over it (using the `didimatch` function defined in `Backend.Vault.Kv`):

```hs
backendNameCodec :: Toml.Key -> Toml.TomlCodec BackendName
backendNameCodec backendNameKey =
  didimatch
    (Just . getBackendName)
    (eitherToMaybe . newBackendName)
    (Toml.text backendNameKey)
```

Reply:
Also:
1. we should move `didimatch` to somewhere where it can be reused, like `Coffer.Util`
2. we should refactor `didimatch` to use `Either Text x` instead of `Maybe x`, so we can report _why_ a conversion failed.

Reply:
Done